Add examples for binary format documentation (#176)
typedefs/typedefs: docs/binary_format/binaryFormat.md

# Typedefs binary format

> Terms are serialised in a type-directed way, given a typedefs

Reviewer: ```suggestion
Terms are serialised in a type-directed way, given a Typedefs
```

description --- terms make no sense on their own. If the type is
unknown, a description of the type could first be serialised, followed

> by the serialisation of the term (this is future work).

Reviewer: ```suggestion
by the serialisation of the term. (This is future work.)
```

Author: It is not clear to me why this change is necessary. Could you explain your reasoning to me?

Reviewer: In this case, it's more or less that the sentence still has to make sense without parens. Alternatively, emdash (or endash, I always forget which is which) could work, e.g. "-- this is future work.", or "(which is future work)."

Reviewer: Oh, I just noticed you do use `---` in other places.

Author: I think the `---` is from fred but I'll put it here as well.

- Terms of type `TSum ts` with `|ts| = 2 + k` are serialised as an integer `i` (with `i < 2 + k`, but this is not exploited), followed by the serialisation of a term of type `ts[i]`.
- Terms of type `TProd ts` with `|ts| = 2 + k` are serialised as the serialisation of `ts[0]`, ..., `ts[1+k]`. This relies on being able to compute the width of each serialised term.
- Terms of type `Tvar j` are serialised with a given "parameter" serialiser --- this is used to serialise `TMu`s.
- Terms of type `TMu ts` with |ts| = k are serialised as an integer `i` (with `i < k`, but this is not exploited), followed by the serialisation of a term of type ts[i], where the "parameter" serialiser for the new type variable is instantiated to this described procedure.
- Terms of type `TApp f xs` with `|xs| = k` are serialised as terms of type `f`, with "parameter" serialisers given by the `k` serialisations of terms of type `xs[k]`.

# Example

Given this typedef:

```
(name Boolean (mu (True 1) (False 1)))

(name Try (+ (var 0) (var 1)))

(name LinkedList (mu (Nil 1) (Cons (* (var 1) (var 0)))))
```
_You can find this definition in the `example.tdef` file_

> We can generate a new Haskell file that will define those type in Haskell and will create serializers and

Reviewer: ```suggestion
We can generate a new Haskell file that will define those types in Haskell and will create serialisers and
```

> deserializers for terms of those types.

Reviewer: ```suggestion
deserialisers for terms of those types.
```

> We generate the haskell code using the typedef command line tool. See the [installation tutorial][INSTALL].

Reviewer: ```suggestion
We generate the haskell code using the `typedefs` command line tool. See the [installation tutorial][INSTALL].
```

```
typedefs -s example.tdef -d example.hs
```

You can find the generated file in this repo at `docs/Binary format/example.hs`<sup>[1](#footnote)</sup>

## Interacting with Haskell

> In order to use typedefs, you need to import its definitions in your codebase.

Reviewer: ```suggestion
In order to use Typedefs, you need to import its definitions in your codebase.
```

> Here is an example of a Haskell file that imports the typedefs defintions and uses them for serializing and

Reviewer: ```suggestion
Here is an example of a Haskell file that imports the Typedefs definitions and uses them for serializing and
```

> deserializing terms that correspond to the specified types:

Reviewer: ```suggestion
deserialising terms that correspond to the specified types:
```

                                         else print "values are inconsistent"
          let encodedList = encodeList listBool
          print $ displayBytes encodedList
          let decodedList = runDeserialiser (decodeLinkedList decodeBoolean) encodedList
          case decodedList of
            Nothing -> print "decoding failed"
            Just (d, _) -> if d == listBool then print "decoding succeeded"
                                            else print "values are inconsistent"
          let encodedTry = Main.encodeTry success
          print $ displayBytes encodedTry
          let decodedTry = runDeserialiser (decodeTry decodeBoolean (decodeLinkedList decodeBoolean)) encodedTry
          case decodedTry of
            Nothing -> print "decoding failed"
            Just (d, _) -> if d == success then print "decoding succeeded"
                                           else print "values are inconsistent"
            

```
_You can find this program in the `main.hs` file_

You can compile this main file with GHC, using the usual

```
ghc main.hs
```

command, and then execute the program using

```
./main
```

> The output shoudl be :

Reviewer: ```suggestion
The output should be :
```

```
"true : 0"
"false : 1"
"1"
Just (False,"")
"decoding succeeded"
"1110110"
"decoding succeeded"
"11110110"
"decoding succeeded"
```

> Here is a table that recaps how those types are being serialized: 

Reviewer: ```suggestion
Here is a table that recaps how those types are being serialised: 
```

| Term                         | Binary    |
|------------------------------|-----------|
| `False`                      | `1`       |
| `True`                       | `0`       |
| `[False, True, False]`       | `111010`  |
| `Right [False, True, False]` | `1111010` |

##### Footnote

> We had to manually modify the generated haskell file in order to fix some scoping issues and add a module header. 

Reviewer: ```suggestion
We had to manually modify the generated Haskell file in order to fix some scoping issues and add a module header. 
```
